### PR TITLE
Relayer service

### PR DESCRIPTION
**Pull request by musnit**
Completed all core functionality for automated relayer

### CODE REVIEW

**Comment by musnit** [comment]
@denalimarsh due to merge issues on this branch, i'm force pushing it onto master and keeping a backup of master as master_backup

we can just close this PR once your review is done

File: README.md [hunk 1]
```diff
@@ -184,9 +158,43 @@ ebcli rest-server --trust-node
...
+ - 6. Enter the following for the variables under function lock():
+  _recipient = [HASHED_COSMOS_RECIPIENT_ADDRESS] *(for testuser cosmos1pjtgu0vau2m52nrykdpztrt887aykue0hq7dfh, enter "0x636f736d6f7331706a74677530766175326d35326e72796b64707a74727438383761796b756530687137646668")*
+  _token = [DEPLOYED_TOKEN_ADDRESS] *(erc20 not currently supported, enter "0x0000000000000000000000000000000000000000" for ethereum)*
+  _amount = [WEI_AMOUNT]
+ - 7. Enter the same number from _amount as the transaction's value (in wei)
+ - 8. Select "transact" to send the lock() transaction
+
 ## Using the modules in other projects
+
+The ethbridge and oracle modules can be used in other cosmos-sdk applications by copying them into your application's modules folders and including them in the same way as in the example application. Each module may be moved to its own repo or integrated into the core Cosmos-SDK in future, for easier usage.
```
Comment (denalimarsh) [inline README.md:197]: Good addition

File: README.md [hunk 2]
```diff
@@ -184,9 +158,43 @@ ebcli rest-server --trust-node
...
+
 There are 2 nuances you need to be aware of when using these modules in other Cosmos-SDK projects.
- - A specific version of golang.org/x/crypto (ie tendermint/crypto) is needed for compatability with go-ethereum. See the Gopkg.toml for constraint details
- - The govendor steps in the hack as above are needed
+ - A specific version of golang.org/x/crypto (ie tendermint/crypto) is needed for compatability with go-ethereum. See the Gopkg.toml for constraint details. There is an open pull request to tendermint/crypto to add compatbility, but until that is merged you need to use the customized version (https://github.com/tendermint/crypto/pull/1)
+ - The govendor steps in the application as above are needed
 
  For instructions on building and deploying the smart contracts, see the README in their folder.
```
Comment (denalimarsh) [inline README.md:203]: It's fine for now, but going forward lets integrate the two so there isn't confusion about the relationship between Peggy and the Bridge modules.

File: README.md [hunk 3]
```diff
@@ -184,9 +158,43 @@ ebcli rest-server --trust-node
...
+With the application set up and the relayer running, you can now use Peggy by sending a lock transaction to the smart contract. You can do this from any Ethereum wallet/client that supports smart contract transactions.
+
+The easiest way to do this for now, assuming you have Metamask setup for Ropsten in the browser is to use remix or mycrypto as the frontend, for example:
+ - 1. Go to remix.ethereum.org
+ - 2. Compile Peggy.sol with solc v0.5.0
+ - 3. Set the environment as Injected Web3 Ropsten
+ - 4. On 'Run' tab, select Peggy and enter "0x3de4ef81Ba6243A60B0a32d3BCeD4173b6EA02bb" in 'At Address' field
+ - 5. Select 'At Address' to load the deployed contract
+ - 6. Enter the following for the variables under function lock():
+  _recipient = [HASHED_COSMOS_RECIPIENT_ADDRESS] *(for testuser cosmos1pjtgu0vau2m52nrykdpztrt887aykue0hq7dfh, enter "0x636f736d6f7331706a74677530766175326d35326e72796b64707a74727438383761796b756530687137646668")*
```
Comment (denalimarsh) [inline README.md:189]: Fine for functioning prototype, but will need to simplify this input for users somehow. We could either do the hash on contract or under the hood in a Peggy cli which allows users to send txs directly to the contract from within terminal.

File: cmd/ebrelayer/txs/relay.go [hunk 4]
```diff
@@ -50,5 +52,20 @@ func RelayEvent(chainId string, cdc *amino.Codec, claim *types.EthBridgeClaim) e
 
 	cliCtx.PrintResponse = true
 
-	return utils.CompleteAndBroadcastTxCLI(txBldr, cliCtx, []sdk.Msg{msg})
```
Comment (denalimarsh) [inline cmd/ebrelayer/txs/relay.go]: Deconstruct the `CompleteAndBroadcastTXCLI()` call so we can automatically apply the validator's passphrase to the tx makes sense to me.

File: cmd/ebrelayer/relayer/relayer.go [hunk 5]
```diff
@@ -32,8 +33,13 @@ import (
 // -------------------------------------------------------------------------
 
 func InitRelayer(cdc *amino.Codec, chainId string, provider string,
-								 contractAddress common.Address, eventSig string,
-								 validator sdk.AccAddress) error {
+	contractAddress common.Address, eventSig string,
+	validator sdk.AccAddress) error {
+
+	passphrase, err := keys.GetPassphrase("validator")
```
Comment (denalimarsh) [inline cmd/ebrelayer/relayer/relayer.go]: Is this method of pulling the validator's key secure for multiple independent users? Can this be applied safely for local validators in addition to the public test "validator" account being queried here?

**Review by denalimarsh — COMMENTED** [review]
Question about production vs. testing querying of validator passphrases.

**EDIT - security concerns are mitigated by other changes in this PR, namely the parameterization of the validator as opposed to utilization of the public test validator's key. Looks good!

File: cmd/ebrelayer/txs/relay.go [hunk 6]
```diff
@@ -40,32 +42,38 @@ func RelayEvent(chainId string, cdc *amino.Codec, passphrase string, claim *type
...
 	//prepare tx
 	txBldr, err = utils.PrepareTxBuilder(txBldr, cliCtx)
 	if err != nil {
+		fmt.Printf("Msg prepare error: %s", err)
 		return err
 	}
 
 	// build and sign the transaction
-	txBytes, err := txBldr.BuildAndSign("validator", passphrase, []sdk.Msg{msg})
+	txBytes, err := txBldr.BuildAndSign(validatorName, passphrase, []sdk.Msg{msg})
```
Comment (denalimarsh) [inline cmd/ebrelayer/txs/relay.go:65]: Ahh, "Validator" is updated to a parameterized validatorName here. This answers my question.

File: cmd/ebrelayer/txs/relay.go [hunk 7]
```diff
@@ -40,32 +42,38 @@ func RelayEvent(chainId string, cdc *amino.Codec, passphrase string, claim *type
 
 	err := cliCtx.EnsureAccountExistsFromAddr(claim.Validator)
 	if err != nil {
-		fmt.Errorf("Validator account error: %s", err)
+		fmt.Printf("Validator account error: %s", err)
```
Comment (denalimarsh) [inline cmd/ebrelayer/txs/relay.go:45]: Good updates to the Error logging

File: README.md (hunk 2 above)
Comment (musnit) [reply README.md:203]: Not sure what you mean/what this has to do with that relationship?

File: README.md (hunk 3 above)
Comment (musnit) [reply README.md:189]: Yeh, this needs to be improved. Or we could do the hash on the relayer, couldn't we?

File: README.md [hunk 8]
```diff
@@ -167,7 +167,10 @@ For automated relaying, there is a relayer service that can be run that will aut
 ebrelayer status
 
 # Initialize the Relayer service for automatic claim processing
-ebrelayer init testing wss://ropsten.infura.io/ws 3de4ef81Ba6243A60B0a32d3BCeD4173b6EA02bb "LogLock(bytes32,address,bytes,address,uint256,uint256)" $(ebcli keys show validator -a)
+ebrelayer init testing wss://ropsten.infura.io/ws 3de4ef81Ba6243A60B0a32d3BCeD4173b6EA02bb "LogLock(bytes32,address,bytes,address,uint256,uint256)" validator
```
Comment (denalimarsh) [inline README.md:170]: We can simplify this command a bit, but for now it's acceptable for the prototype.

File: cmd/ebrelayer/relayer/relayer.go (hunk 5 above)
Comment (musnit) [reply cmd/ebrelayer/relayer/relayer.go]: You're reviewing an outdated file?

File: cosmos-ethereum-bridge.postman_collection.json [hunk 9]
```diff
@@ -96,46 +103,54 @@
...
+						"key": "Content-Type",
+						"name": "Content-Type",
+						"value": "application/json",
+						"type": "text"
+					}
+				],
 				"body": {
 					"mode": "raw",
-					"raw": ""
+					"raw": "{\n    \"base_req\": {\n        \"chain_id\": \"testing\",\n        \"from\": \"cosmos18hf69vxn8a3tkladruxgxgv8tl8sl54gygdh29\"\n    },\n    \"nonce\": \"0\",\n    \"ethereum_sender\": \"0x7B95B6EC7EbD73572298cEf32Bb54FA408207359\",\n    \"amount\": \"4eth\",\n    \"cosmos_receiver\": \"cosmos19l0hyjpzm8xkwlu84my4f0npd2ranxt2yfztux\"\n}"
```
Comment (denalimarsh) [inline cosmos-ethereum-bridge.postman_collection.json:135]: +1 looks good

**Review by denalimarsh — COMMENTED** [review]
These changes satisfy the question I had about the validator passphrase security, as the validatorName is parameterized and "validator" isn't used directly to pull the public test validator's key for signatures.

**Pull request merged**